photon-ml/src/main/scala/com/linkedin/photon/ml/function/DiffFunction.scala

```diff
@@ -15,307 +15,38 @@
-      func: DiffFunction[Datum],
-      regularizationContext: RegularizationContext,
-      regWeight: Double): DiffFunction[Datum] = {
-
-    val (l1Weight, l2Weight) = (
-      regularizationContext.getL1RegularizationWeight(regWeight),
-      regularizationContext.getL2RegularizationWeight(regWeight))
-
-    (l1Weight,  l2Weight) match {
-      case (0.0, 0.0) =>
-        // No regularization
-        func
-      case (0.0, l2) =>
-        // Only l2 regularization
-        withL2Regularization(func, l2)
-      case (l1, 0.0) =>
-        // Only l1 regularization
-        withL1Regularization(func, l1)
-      case (l1, l2) =>
-        // L1 and L2 regularization
-        val funcWithL2 = withL2Regularization(func, l2)
-        withL1Regularization(funcWithL2, l1)
-    }
-  }
+  * Trait for differentiable function. The value/gradient of a differentiable function depends on the model coefficients
+  * (e.g., the coefficients learned in a generalized linear model).
+  *
+  * FAQ: Why do the functions take an ObjectProvider[ Vector[Double] ] for coefficients instead of a Vector[Double]?
+  *      Why may the functions potentially receive a Broadcast[ Vector[Double] ]?
+  * A: In the optimization procedure, the gradient/value/calculate function might be called repeatedly several times
+  *    given the same input coefficients (e.g., during conjugate gradient method). As a result, if the input
+  *    coefficients type is Vector, we need to re-broadcast the input vector every time to make it accessible in the
+  *    function, even when the value of the input vector itself is unchanged. This would potentially cause network
+  *    I/O problems if the functions are called frequently, and would result in a waste of computation as well.
+  *    However, if the input coefficients are of type BroadCast[ Vector[Double] ], we can re-use them for different
+  *    function calls as long as the underlying vector hasn't changed. Since GAME random effect model computations
+  *    are local to a single node, these functions may receive either a broadcast Vector as described above, or a
+  *    regular Vector.
+  */
+trait DiffFunction extends ObjectiveFunction {
+  /**
+    * Compute the gradient of the function given model coefficients.
+    *
+    * @param coefficients The model coefficients used to compute the function's gradient
+    * @return The computed gradient of the function
+    */
+  protected[ml] def gradient(coefficients: Vector[Double]): Vector[Double]
+
+  /**
+    * Compute both the value and the gradient of the function given model coefficients (computing value and gradient at
+    * once is sometimes more efficient than computing them sequentially).
+    *
+    * @param coefficients The model coefficients used to compute the function's value and gradient
+    * @return The computed value and gradient of the function
+    */
+  protected[ml] def calculate(coefficients: Vector[Double]): (Double, Vector[Double])
```

Comment:
@joshvfleming 
DiffFunction is just an interface now, all of the bulk has been stripped out (same goes for TwiceDiffFunction).